src/java.base/share/classes/java/util/Formatter.java

```diff
@@ -2008,15 +2008,28 @@
  * @since 1.5
  */
 public final class Formatter implements Closeable, Flushable {
+
+    // Caching DecimalFormatSymbols
+    static DecimalFormatSymbols DFS = null;
+    static DecimalFormatSymbols getDecimalFormatSymbols(Locale locale) {
+        DecimalFormatSymbols dfs = DFS;
+        if (dfs != null && dfs.getLocale() == locale) {
+            return dfs;
+        }
+        dfs = DecimalFormatSymbols.getInstance(locale);
+        DFS = dfs;
+        return dfs;
+    }
+
+    // Caching zero.
+    static char getZero(Locale locale) {
+        return locale == null ? '0' : getDecimalFormatSymbols(locale).getZeroDigit();
```

Comment:
While we are at it, it would be beneficial to cache locale-dependent grouping and decimal separators too.

Reply:
Noted